docs/cookbook/recipe_sonata_admin_without_user_bundle.rst

```diff
@@ -0,0 +1,451 @@
+                Security::LAST_USERNAME,
+                $data['email']
+            );
+
+            return $data;
+        }
+
+        public function getUser($credentials, UserProviderInterface $userProvider)
+        {
+            return $userProvider->loadUserByUsername($credentials['email']);
+        }
+
+        public function checkCredentials($credentials, UserInterface $user)
+        {
+            if (!$this->passwordEncoder->isPasswordValid($user, $credentials['password'])) {
+                return false;
+            }
+
+            if (!$user->hasRole('ROLE_ADMIN')) {
+                throw new CustomUserMessageAuthenticationException("You don't have permission to access that page.");
+            }
+
+            return true;
+        }
+
+        public function onAuthenticationFailure(Request $request, AuthenticationException $exception)
+        {
+            $request->getSession()->set(Security::AUTHENTICATION_ERROR, $exception);
+            return new RedirectResponse($this->router->generate('admin_login'));
+        }
+
+        protected function getLoginUrl()
+        {
+            return new RedirectResponse($this->router->generate('admin_login'));
+        }
+
+        public function onAuthenticationSuccess(Request $request, TokenInterface $token, $providerKey)
+        {
+            return new RedirectResponse($this->router->generate('sonata_admin_dashboard'));
+        }
+    }
+
+LoginController
+^^^^^^^^^^^^^^^
+
+A Controller, used to render login form. Logout is left empty intentionally because
+this will be handled by Symfony, but we still need to register that route.
+
+.. code-block:: php
+
+    namespace App\Controller;
+
+    use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+    use App\Form\LoginForm;
+    use Symfony\Component\Routing\Annotation\Route;
+    use Symfony\Component\Security\Http\Authentication\AuthenticationUtils;
+
+    class LoginController extends Controller
+    {
+        /**
+         * @var AuthenticationUtils
+         */
+        private $authenticationUtils;
+
+        public function __construct(AuthenticationUtils $authenticationUtils)
+        {
+            $this->authenticationUtils = $authenticationUtils;
+        }
+
+        /**
+         * @Route("/admin/login", name="admin_login")
+         */
+        public function loginAction()
+        {
+            $form = $this->createForm(LoginForm::class, [
+                'email' => $this->authenticationUtils->getLastUsername()
+            ]);
+
+            return $this->render('security/login.html.twig', [
+                'last_username' => $this->authenticationUtils->getLastUsername(),
+                'form' => $form->createView(),
+                'error' => $this->authenticationUtils->getLastAuthenticationError(),
+            ]);
+        }
+
+        /**
+         * @Route("/admin/logout", name="admin_logout")
+         */
+        public function logoutAction()
+        {
```

Comment:
Imo this paragraph:
```
A Controller, used to render login form. Logout is left empty intentionally because
this will be handled by Symfony, but we still need to register that route.
```

should be added as a comment in the method body, otherwise its overwritten maybe? but then its copy pasted...hmmm 🤔

Reply:
I can copy a part of it and leave it between `{}`, I guess that should avoid any confusion?

Reply:
overwritte -> not read...